Update Scaling out Airflow Guide

kentdanas opened the pull request:
Updated Scaling out Airflow guide to fix some outdated info and add content based on feedback from solutions and CRE.

@collinmcnulty @dylanbstorey if you have a chance and want to give this a read and make sure I got it all right, that would be much appreciated!

Closes #308, closes #316

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+Airflow has many parameters that impact scaling as your number of DAGs and tasks grows. Tuning these settings can impact DAG parsing and task scheduling performance, parallelism in your Airflow environment, and more. 
 
-To check current values for an existing Airflow environment, go to **Admin** > **Configurations** in the Airflow UI. For more information, read [Setting Configuration Options](https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html) or [Configuration Reference](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html) in Airflow documentation.
+The reason Airflow has so many nobs at different levels is that, as an agnostic orchestrator, Airflow is used for a wide variety of use cases and by a wide variety of personas. Airflow admins or devops engineers might look to tune scaling parameters at the environment level to ensure their supporting infrastructure isn't overstressed, while DAG authors might look to tune scaling parameters at the ...
```
spelling: knobs, DevOps

kentdanas replied:
Ha, "nobs" was a funny typo in this context. Thanks for catching

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+Core settings control the number of processes running concurrently and how long those processes can run. The associated environment variables for all parameters in this section are of the form `AIRFLOW__CORE__PARAMETER_NAME`.
 
-  If you increase the amount of resources available to Airflow (such as Celery Workers or Kubernetes Pods) and notice that tasks are still not running as expected, you might have to increase the values of both `parallelism` and `dag_concurrency`.
+- **`parallelism`:** Default 32. This is the maximum number of tasks that can run concurrently within a single Airflow environment. If this setting is set to 32, no more than 32 tasks can run at once across all DAGs. Think of this as "maximum active tasks anywhere." If you notice that tasks are stuck in a queue for extended periods of time, this is a value you may want to increase.
```
The state is called "Queued", so that might be clearer wording than "in a queue"

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+- **`parallelism`:** Default 32. This is the maximum number of tasks that can run concurrently within a single Airflow environment. If this setting is set to 32, no more than 32 tasks can run at once across all DAGs. Think of this as "maximum active tasks anywhere." If you notice that tasks are stuck in a queue for extended periods of time, this is a value you may want to increase.
 
-- **`worker_concurrency`** is only applicable to users running the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html). It determines how many tasks each Celery Worker can run at any given time. If this value is not set, the Celery Executor will run a maximum of 16 tasks concurrently by default.
+- **`max_active_tasks_per_dag`:** (formerly `dag_concurrency`) Default 16. This determines how many task instances the Airflow Scheduler is able to schedule at once per DAG. Think of this as "maximum tasks that can be scheduled at once, per DAG."
```
Use this setting to prevent any one DAG from taking up too many of the available slots from parallelism or your pools. This helps DAGs to be good neighbors to one another.

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+- **`max_active_tasks_per_dag`:** (formerly `dag_concurrency`) Default 16. This determines how many task instances the Airflow Scheduler is able to schedule at once per DAG. Think of this as "maximum tasks that can be scheduled at once, per DAG."
 
-  This number is limited by `dag_concurrency`. If you have 1 Worker and want it to match your environment's capacity, `worker_concurrency` should be equal to `parallelism`. If you increase `worker_concurrency`, you might also need to provision additional CPU and/or memory for your Workers.   
+  If you increase the amount of resources available to Airflow (such as Celery Workers or Kubernetes Pods) and notice that tasks are still not running as expected, you might have to increase the values of both `parallelism` and `max_active_tasks_per_dag`.
```
Kubernetes Resources, not Kubernetes pods. In K8s you add available memory and CPU and Airflow eats that to form pods.

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+  If you increase the amount of resources available to Airflow (such as Celery Workers or Kubernetes Pods) and notice that tasks are still not running as expected, you might have to increase the values of both `parallelism` and `max_active_tasks_per_dag`.
 
-- **`parsing_processes`** is the maximum number of threads that can run on the Airflow Scheduler at once. Increasing the value of `parallelism` results in additional strain on the Scheduler, so we recommend increasing `parsing_processes` as well. If you notice a delay in tasks being scheduled, you might need to increase this value or provision additional resources for your Scheduler. If you're ru ...
+- **`max_active_runs_per_dag`:** Default 16. This determines the maximum number of active DAG Runs (per DAG) the Airflow Scheduler can create at any given time. In Airflow, a [DAG Run](https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html) represents an instantiation of a DAG in time, much like a task instance represents an instantiation of a task.
```
This parameter is most relevant when backfilling or if Airflow has to catch up from missed DAG Runs. Consider how you want to handle these scenarios when setting this parameter.

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+- **`max_active_runs_per_dag`:** Default 16. This determines the maximum number of active DAG Runs (per DAG) the Airflow Scheduler can create at any given time. In Airflow, a [DAG Run](https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html) represents an instantiation of a DAG in time, much like a task instance represents an instantiation of a task.
 
-    > **Note:** This setting was renamed in Airflow 1.10.14. In earlier versions, it is defined as `max_threads` ([source](https://github.com/apache/airflow/commit/486134426bf2cd54fae1f75d9bd50715b8369ca1)).
+- **`dag_file_processor_timeout`:**: Default 30 seconds. This is how long a `DagFileProcessor`, which processes a DAG file, can run before timing out. If your DAG processing logs show timeouts, try increasing this value.
```
If your DAG processing logs show timeouts or if your DAG is not showing up either in the list of DAGs or the import errors ...

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,79 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
-- **`concurrency`** is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `dag_concurrency` (described above) is assumed.
+  If are utilizing `catchup` or `backfill` for your DAG, consider defining this parameter to ensure you don't accidentally trigger a high number of DAG runs.
+- **`max_active_tasks`:** This is the total number of tasks that can run at the same time for a given DAG run. It essentially controls the parallelism within your DAG. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
+- **`concurrency`:** This is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `dag_concurrency` (described above) is assumed.
```
The value of max_active_tasks_per_dag is assumed

collinmcnulty replied:
Is it still called concurrency here?

kentdanas replied:
Yep, still "concurrency" in the 2.2 docs https://airflow.apache.org/docs/apache-airflow/stable/faq.html#scheduling-dag-file-parsing

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,79 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+There are two primary task-level Airflow settings users can define in code:
 
-- **`task_concurrency`** is a limit to the amount of times the same task can execute across multiple DAG Runs.
+- **`max_active_tis_per_dag`:** Formerly `task_concurrency`. This is a limit to the amount of times the same task can run concurrently across all DAG Runs.
```
For instance, if a task reaches out to an external resource such as a data table that you do not want multiple tasks modifying at once, you can set this value to 1.

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,79 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+You can also tune your **`worker_concurrency`** (environment variable `AIRFLOW__CELERY__WORKER_CONCURRENCY`), which determines how many tasks each Celery worker can run at any given time. If this value is not set, the Celery executor will run a maximum of 16 tasks concurrently by default.
 
-3. Assign your tasks to the pool:
+Your `worker_concurrency` is limited by `max_active_tasks_per_dag`. If you increase `worker_concurrency`, you might also need to provision additional CPU and/or memory for your workers.   
```
I don't think that's true. There's no reason you couldn't run 100 workers each of which can only handle a worker_concurrency of 2 (that I know of).

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,79 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+    - Try changing: `scheduler_heartbeat_sec`, `parsing_processes`
+- Issue: An individual DAG is having trouble running tasks in parallel, while other DAGs seem unaffected
+    - Potential cause: Possible DAG-level bottleneck
+    - Try changing: `max_active_task_per_dag`
```
missing s in tasks

Try looking at pools. See if you're at your overall parallelism limit.

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,79 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+    - Try changing: `worker_concurrency` (if using Celery), `parallelism`
+- Issue: DAGs are stuck in queued state, but not running
+    - Potential cause: The number of tasks being scheduled may be beyond the capacity of your Airflow infrastructure.
+    - Try changing: `scheduler_heartbeat_sec`, `parsing_processes`
```
The try changing here is incorrect. By the time it gets to queued it's past parsing and the scheduler. If using K8s executor: check that there are available resources in the namespace and check if worker_pods_creation_batch_size can be increased. If using Celery check if worker_concurrency can be increased (and generally if you have workers available to do the work).

kentdanas commented:
@collinmcnulty thanks for your comments! I think I addressed everything, but let me know if you catch anything else.

collinmcnulty commented:
@wolfier Curious as to your thoughts on this too

wolfier commented:
Great work! This will be immensely helpful to pass out to customer :D

kentdanas commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+- **`max_active_runs_per_dag`:** Default 16. This determines the maximum number of active DAG Runs (per DAG) the Airflow Scheduler can create at any given time. In Airflow, a [DAG Run](https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html) represents an instantiation of a DAG in time, much like a task instance represents an instantiation of a task. This parameter is most relevant whe ...
 
-    > **Note:** This setting was renamed in Airflow 1.10.14. In earlier versions, it is defined as `max_threads` ([source](https://github.com/apache/airflow/commit/486134426bf2cd54fae1f75d9bd50715b8369ca1)).
+- **`dag_file_processor_timeout`:**: Default 30 seconds. This is how long a `DagFileProcessor`, which processes a DAG file, can run before timing out. If your DAG processing logs show timeouts, or if your DAG is not showing up either in the list of DAGs or the import errors, try increasing this value.
```
Good catch, thanks

collinmcnulty commented on guides/airflow-scaling-workers.md:
```diff
@@ -61,7 +61,7 @@ Scheduler settings control how the scheduler parses DAG files and creates DAG ru
   It is helpful know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you may see performance issues.
 
-- **`parsing_processes`:** (formerly `max_threads`) Default 2. The scheduler can run multiple processes in parallel to parse dags. This defines how many processes will run. Increasing this value can help serialize DAGs more efficiently if you have a large number of them. Note that if you are running multiple schedulers, this value will apply to *each* of them.
+- **`parsing_processes`:** (formerly `max_threads`) Default 2 x vCPU. The scheduler can run multiple processes in parallel to parse dags. This defines how many processes will run. Increasing this value can help serialize DAGs more efficiently if you have a large number of them. Note that if you are running multiple schedulers, this value will apply to *each* of them.
```
The default really is 2. We *recommend* you set it to 2 x vCPU

kentdanas replied:
Sigh, moving too quickly. Thanks for catching!

kentdanas commented:
@jwitz this should be good to go content-wise. If I could get a copy review whenever it's convenient that would be much appreciated!

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`dag_dir_list_interval`:** Default 300 seconds (5 minutes). This is how often to scan the DAGs directory for new files. The lower the number, the faster new DAGs will be processed but the higher your CPU usage.
+
+  It is helpful know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you may see performance issues.
```
```suggestion
  It is helpful know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you might see performance issues.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+  It is helpful know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you may see performance issues.
+
+- **`parsing_processes`:** (formerly `max_threads`) Default 2. The scheduler can run multiple processes in parallel to parse dags. This setting defines how many processes will run. We recommend setting a value of 2 x vCPU. Increasing this value can help serialize DAGs more efficiently if you have a large number of them. Note that if you are running multiple schedulers, this value will apply to *e ...
```
```suggestion
- **`parsing_processes`:** (formerly `max_threads`) The scheduler can run multiple processes in parallel to parse DAGs. This setting defines how many processes can run in parallel. We recommend setting a value of 2x your available vCPUs. Increasing this value can help serialize DAGs more efficiently if you have a large number of them. Note that if you are running multiple schedulers, this value will apply to *each* of them. By default, this value is set to 2. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`parsing_processes`:** (formerly `max_threads`) Default 2. The scheduler can run multiple processes in parallel to parse dags. This setting defines how many processes will run. We recommend setting a value of 2 x vCPU. Increasing this value can help serialize DAGs more efficiently if you have a large number of them. Note that if you are running multiple schedulers, this value will apply to *e ...
+
+- **`file_parsing_sort_mode`**: Default `modified_time`. This determines how the scheduler will list and sort DAG files to decide the parsing order. Set to one of: `modified_time`, `random_seeded_by_host` and `alphabetical`.
```
```suggestion
- **`file_parsing_sort_mode`:** This determines how the scheduler will list and sort DAG files to decide the parsing order. Set to one of: `modified_time`, `random_seeded_by_host` and `alphabetical`. The default setting is `modified_time`. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`file_parsing_sort_mode`**: Default `modified_time`. This determines how the scheduler will list and sort DAG files to decide the parsing order. Set to one of: `modified_time`, `random_seeded_by_host` and `alphabetical`.
+
+- **`scheduler_heartbeat_sec`:** Default 5 seconds. The scheduler constantly tries to trigger new tasks. This defines how often the scheduler should run (in seconds).
```
```suggestion
- **`scheduler_heartbeat_sec`:** This setting defines how often the scheduler should run (in seconds) to trigger new tasks. The default value is 5 seconds. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`scheduler_heartbeat_sec`:** Default 5 seconds. The scheduler constantly tries to trigger new tasks. This defines how often the scheduler should run (in seconds).
+
+- **`max_dagruns_to_create_per_loop`:** Default 10. This is the max number of DAGs to create DagRuns for per scheduler loop. You can use this option to free up resources for scheduling tasks by **decreasing** the value.
```
```suggestion
- **`max_dagruns_to_create_per_loop`:** This is the max number of DAGs to create Dag runs for per scheduler loop. You can use this option to free up resources for scheduling tasks by decreasing the value. The default value is 10 seconds. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`max_dagruns_to_create_per_loop`:** Default 10. This is the max number of DAGs to create DagRuns for per scheduler loop. You can use this option to free up resources for scheduling tasks by **decreasing** the value.
+
+- **`max_tis_per_query`:** Default 512. This parameter changes the batch size of queries to the metastore in the main scheduling loop. If the value is higher, you can process more `tis` per query, but your query may become too complex and create a performance bottleneck. 
```
```suggestion
- **`max_tis_per_query`:** This parameter changes the batch size of queries to the metastore in the main scheduling loop. If the value is higher, you can process more `tis` per query, but your query may become too complex and create a performance bottleneck. The default value is 512 queries. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+### DAG-level Airflow Settings
 
-There are two primary DAG-level Airflow settings users can define in code:
+DAG-level settings apply only to specific DAGs and are defined in your DAG code. You should look to modify DAG-level settings if you want to performance tune a particular DAG, especially in cases where that DAG is hitting an external system (e.g. an API or database) that might cause performance issues if hit too frequently. In general, DAG-level settings will supersede environment-level settings  ...
```
```suggestion
DAG-level settings apply only to specific DAGs and are defined in your DAG code. You should modify DAG-level settings if you want to performance tune a particular DAG, especially in cases where that DAG is hitting an external system (e.g. an API or database) that might cause performance issues if hit too frequently. When a there's a setting at both the DAG-level and environment-level, the DAG-level setting takes precedence.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+DAG-level settings apply only to specific DAGs and are defined in your DAG code. You should look to modify DAG-level settings if you want to performance tune a particular DAG, especially in cases where that DAG is hitting an external system (e.g. an API or database) that might cause performance issues if hit too frequently. In general, DAG-level settings will supersede environment-level settings  ...
 
-- **`max_active_runs`** is the maximum number of active DAG Runs allowed for the DAG in question. Once this limit is hit, the Scheduler will not create new active DAG Runs. If this setting is not defined, the value of `max_active_runs_per_dag` (described above) is assumed.
+There are three primary DAG-level Airflow settings users can define in code:
```
```suggestion
There are three primary DAG-level Airflow settings that users can define in code:
```

jwitz commented on guides/airflow-scaling-workers.md:
````diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
   ```
 
-- **`concurrency`** is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `dag_concurrency` (described above) is assumed.
+  If are utilizing `catchup` or `backfill` for your DAG, consider defining this parameter to ensure you don't accidentally trigger a high number of DAG runs.
````
```suggestion
  If are utilizing `catchup` or `backfill` for your DAG, consider defining this parameter to ensure that you don't accidentally trigger a high number of DAG runs.
```

jwitz commented on guides/airflow-scaling-workers.md:
````diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+There are three primary DAG-level Airflow settings users can define in code:
 
-  ```
+- **`max_active_runs`:** This is the maximum number of active DAG Runs allowed for the DAG in question. Once this limit is hit, the Scheduler will not create new active DAG Runs. If this setting is not defined, the value of `max_active_runs_per_dag` (described above) is assumed.
````
```suggestion
- **`max_active_runs`:** This is the maximum number of active DAG Runs allowed for the DAG in question. Once this limit is hit, the Scheduler will not create new active DAG Runs. If this setting is not defined, the value of the environment-level setting `max_active_runs_per_dag` is assumed.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
 
-- **`concurrency`** is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `dag_concurrency` (described above) is assumed.
+  If are utilizing `catchup` or `backfill` for your DAG, consider defining this parameter to ensure you don't accidentally trigger a high number of DAG runs.
+- **`max_active_tasks`:** This is the total number of tasks that can run at the same time for a given DAG run. It essentially controls the parallelism within your DAG. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
```
```suggestion
- **`max_active_tasks`:** This is the total number of tasks that can run at the same time for a given DAG run. It essentially controls the parallelism within your DAG. If this setting is not defined, the value of the environment-level setting `max_active_tasks_per_dag` is assumed.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
-- **`concurrency`** is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `dag_concurrency` (described above) is assumed.
+  If are utilizing `catchup` or `backfill` for your DAG, consider defining this parameter to ensure you don't accidentally trigger a high number of DAG runs.
+- **`max_active_tasks`:** This is the total number of tasks that can run at the same time for a given DAG run. It essentially controls the parallelism within your DAG. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
+- **`concurrency`:** This is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
```
```suggestion
- **`concurrency`:** This is the maximum number of task instances allowed to run concurrently across all active DAG runs for a given DAG. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of the environment-level setting `max_active_tasks_per_dag` is assumed.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+- **`max_active_tasks`:** This is the total number of tasks that can run at the same time for a given DAG run. It essentially controls the parallelism within your DAG. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
+- **`concurrency`:** This is the maximum number of task instances allowed to run concurrently across all active DAG runs of the DAG for which this setting is defined. This allows you to set 1 DAG to be able to run 32 tasks at once, while another DAG might only be able to run 16 tasks at once. If this setting is not defined, the value of `max_active_tasks_per_dag` (described above) is assumed.
 
   For example:
```
Why do we have an example just for this setting? I think it's save to remove this, unless it's describing how to set DAG-level settings in general (and in that case, let's say that)

kentdanas replied:
Yeah this was legacy and I don't think it's needed with the way we rewrote. Will remove it

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+### Task-level Airflow Settings
 
-There are two primary task-level Airflow settings users can define in code:
+Task-level settings can be used to implement even finer grain control within a specific DAG(s) and are defined in your operators. You should look to modify task-level settings if you have specific types of tasks that are known to cause performance issues. 
```
```suggestion
Task-level settings are defined in a task's operators and can be used to implement even finer-grained control of performance. You should modify task-level settings if you have specific types of tasks that are known to cause performance issues. 
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+There are two primary task-level Airflow settings users can define in code:
 
-- **`task_concurrency`** is a limit to the amount of times the same task can execute across multiple DAG Runs.
+- **`max_active_tis_per_dag`:** Formerly `task_concurrency`. This is a limit to the amount of times the same task can run concurrently across all DAG Runs. For instance, if a task reaches out to an external resource such as a data table that you do not want multiple tasks modifying at once, you can set this value to 1.
```
```suggestion
- **`max_active_tis_per_dag`:** (Formerly `task_concurrency`) This the maximum amount of times that the same task can run concurrently across all DAG Runs. For instance, if a task reaches out to an external resource, such as a data table, that should not me modified by multiple tasks at once, then you can set this value to 1.
```

jwitz commented on guides/airflow-scaling-workers.md:
````diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
-  For example, you might set the following in your task definition:
+  You can set the following in your task definition:
 
-  ```
-  t1 = PythonOperator(pool='my_custom_pool', task_concurrency=14)
+  ```python
+  t1 = PythonOperator(pool='my_custom_pool', max_active_tis_per_dag=14)
   ```
````
Why don't we just include a general example of task configuration towards the end of this section? I don't have any further understanding of this setting from seeing this snippet.

jwitz commented on guides/airflow-scaling-workers.md:
````diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
   ```
 
-### Airflow Pools Best Practices
+- **`pool`:** Pools are a way to limit the number of concurrent instances of an arbitrary group of tasks. This is useful if you have a lot of workers or DAG runs in parallel, but you want to avoid an API rate limit or otherwise don't want to overwhelm a data source or destination. For more information, read our [Airflow Pools Guide](https://www.astronomer.io/guides/airflow-pools).
````
```suggestion
- **`pool`:** This setting defines the amount of pools available for a task. Pools are a way to limit the number of concurrent instances of an arbitrary group of tasks. This setting is useful if you have a lot of workers or DAG runs in parallel, but you want to avoid an API rate limit or otherwise don't want to overwhelm a data source or destination. For more information, read our [Airflow Pools Guide](https://www.astronomer.io/guides/airflow-pools).
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+The [Celery executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) utilizes standing workers to run tasks. Scaling with the Celery executor involves choosing both the number and size of the workers available to Airflow. The more workers you have available in your environment, or the larger your workers are, the more capacity you have to run tasks concurrently.
 
-    ![image](https://assets2.astronomer.io/main/guides/airflow-scaling-workers/create_pool.png)
+You can also tune your **`worker_concurrency`** (environment variable `AIRFLOW__CELERY__WORKER_CONCURRENCY`), which determines how many tasks each Celery worker can run at any given time. If this value is not set, the Celery executor will run a maximum of 16 tasks concurrently by default. If you increase `worker_concurrency`, you might also need to provision additional CPU and/or memory for your  ...
```
```suggestion
You can also tune your **`worker_concurrency`** (environment variable: `AIRFLOW__CELERY__WORKER_CONCURRENCY`), which determines how many tasks each Celery worker can run at any given time. By default, the Celery executor will run a maximum of 16 tasks concurrently. If you increase `worker_concurrency`, you might also need to provision additional CPU and/or memory for your workers.   
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+## Executors and Scaling
 
-To create a pool:
+Depending on which executor you choose for your Airflow environment, there are a couple of additional concepts to keep in mind when scaling.
```
```suggestion
Depending on which executor you choose for your Airflow environment, there are a couple of additional settings to keep in mind when scaling.
```

jwitz commented on guides/airflow-scaling-workers.md:
````diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
-  ```
+The [Kubernetes executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/kubernetes.html) launches a pod in a Kubernetes cluster for each task. Since each task runs in its own pod, resources can be specified on an individual task level.
+
+When performance tuning with the Kubernetes executor, it is important to keep in mind the supporting infrastructure of your Kubernetes cluster. Many users will enable auto-scaling on their cluster to ensure they get the benefit of Kubernetes' elasticity.
````
```suggestion
When performance tuning with the Kubernetes executor, it is important to consider the supporting infrastructure of your Kubernetes cluster. Many users will enable auto-scaling on their cluster to ensure they get the benefit of Kubernetes' elasticity.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+
+When performance tuning with the Kubernetes executor, it is important to keep in mind the supporting infrastructure of your Kubernetes cluster. Many users will enable auto-scaling on their cluster to ensure they get the benefit of Kubernetes' elasticity.
+
+You can also tune your **`worker_pods_creation_batch_size`** (environment variable `AIRFLOW__KUBERNETES__WORKER_PODS_CREATION_BATCH_SIZE`), which determines how many pods can be created per scheduler loop. The default is 1, but most users will want to increase this number for better performance, especially if you have concurrent tasks. How high you can increase the value depends on the tolerance  ...
```
```suggestion
You can also tune your **`worker_pods_creation_batch_size`** (environment variable: `AIRFLOW__KUBERNETES__WORKER_PODS_CREATION_BATCH_SIZE`), which determines how many pods can be created per scheduler loop. The default is 1, but most users will want to increase this number for better performance, especially if you have concurrent tasks. How high you can increase the value depends on the tolerance of your Kubernetes cluster.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+
+Scaling your Airflow environment can be more of an art than a science, and is highly dependent on your supporting infrastructure and your DAGs. There are too many potential scaling issues to address them all here, but below are some commonly encountered issues and possible parameters to change. 
+
+- Issue: Tasks scheduling latency is high
```
Let's bold everything to the left of the colons. Or maybe even as a table?

kentdanas replied:
Tables are really tough in markdown, but bolding is a good idea.

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,77 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+
+## Potential Scaling Issues
+
+Scaling your Airflow environment can be more of an art than a science, and is highly dependent on your supporting infrastructure and your DAGs. There are too many potential scaling issues to address them all here, but below are some commonly encountered issues and possible parameters to change. 
```
```suggestion
Scaling your Airflow environment can be more of an art than a science, and it's highly dependent on your supporting infrastructure and your DAGs. There are too many potential scaling issues to address them all here, but below are some commonly encountered issues and possible parameters to change. 
```

jwitz reviewed (changes requested):
Sorry for loads of suggestions, but most are very minor! Only things I'm requesting changes on are the code snippets for DAG and task-level settings. I think we should remove our specific examples and instead only include snippets if we're explaining how to define settings more generally.

kentdanas commented:
Thanks for the review @jwitz! For the code snippets, I removed some of them, but left one each in the DAG-level and task-level settings. I want it to be clear that these settings are defined in DAG code instead of with environment variables like the environment-settings, and I think the examples are helpful to show *how* you can modify them (we're not advising on what they *should* be). I cleaned up the language and changed where they are in the section to hopefully make that clearer. Let me know what you think.

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+> **Note:** If you're running Airflow on Astronomer, you should modify these parameters via Astronomer environment variables. For more information, read [Environment Variables on Astronomer](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables).
 
-### Setting Descriptions
+You should modify environment-level settings if you want to tune performance across *all* of the DAGs in your Airflow environment. This is particularly relevant if you want your DAGs can run adequately on your supporting infrastructure. Specifically, the following settings are relevant:
```
```suggestion
You should modify environment-level settings if you want to tune performance across *all* of the DAGs in your Airflow environment. This is particularly relevant if you want your DAGs to run adequately on your supporting infrastructure. Specifically, the following settings are relevant:
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+- **`parallelism`:** This is the maximum number of tasks that can run concurrently within a single Airflow environment. For example, if this setting is set to 32 then no more than 32 tasks can run at once across all DAGs. Think of this as "maximum active tasks anywhere." If you notice that tasks are stuck queued for extended periods of time, this is a value you may want to increase. By default, t ...
 
-- **`worker_concurrency`** is only applicable to users running the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html). It determines how many tasks each Celery Worker can run at any given time. If this value is not set, the Celery Executor will run a maximum of 16 tasks concurrently by default.
+- **`max_active_tasks_per_dag`:** (formerly `dag_concurrency`) This determines maximum number of tasks that can be scheduled at once, per DAG." Use this setting to prevent any one DAG from taking up too many of the available slots from parallelism or your pools, which helps DAGs be good neighbors to one another. By default, this is set to 16.
```
```suggestion
- **`max_active_tasks_per_dag`:** (formerly `dag_concurrency`) This determines the maximum number of tasks that can be scheduled at once, per DAG." Use this setting to prevent any one DAG from taking up too many of the available slots from parallelism or your pools, which helps DAGs be good neighbors to one another. By default, this is set to 16.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -1,92 +1,75 @@
+
+- **`dag_dir_list_interval`:** This is how often to scan the DAGs directory for new files, in seconds. The lower the value, the faster new DAGs will be processed but the higher your CPU usage. By default, this is set to 300 seconds (5 minutes). 
+
+  It is helpful know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you might see performance issues.
```
```suggestion
  It is helpful to know how long it takes to parse your DAGs (`dag_processing.total_parse_time`) to know what values to choose for `min_file_process_interval` and `dag_dir_list_interval`. If your `dag_dir_list_interval` is less than the amount of time it takes to parse each DAG, you might see performance issues.
```

jwitz commented on guides/airflow-scaling-workers.md:
```diff
@@ -96,62 +79,71 @@ If all of these tasks exist within a single DAG and `dag_concurrency=16`, howeve
+### DAG-level Airflow Settings
 
-There are two primary DAG-level Airflow settings users can define in code:
+DAG-level settings apply only to specific DAGs and are defined in your DAG code. You should modify DAG-level settings if you want to performance tune a particular DAG, especially in cases where that DAG is hitting an external system (e.g. an API or database) that might cause performance issues if hit too frequently. When a there's a setting at both the DAG-level and environment-level, the DAG-lev ...
```
```suggestion
DAG-level settings apply only to specific DAGs and are defined in your DAG code. You should modify DAG-level settings if you want to performance tune a particular DAG, especially in cases where that DAG is hitting an external system (e.g. an API or database) that might cause performance issues if hit too frequently. When a setting exists at both the DAG-level and environment-level, the DAG-level setting takes precedence.
```

jwitz reviewed (approved):
New example settings look great! Some final typos here (probably from my own suggestions haha) but otherwise LGTM
